Add cancellation token to web socket

BartoszChrostowski opened the pull request:
#2057

@Botan626

Botan626 commented:
@BartoszChrostowski Tell me when it's ready for a test.

BartoszChrostowski commented:
It is

Botan626 commented:
Then I'm afraid the issue is still valid. Restarted ASF via restart button in ASF-ui with Log page opened.

Log: https://pastebin.com/raw/eW32aEtS

BartoszChrostowski commented:
Ok, thank you for your time 😊

The behavior changed. 
now the exception type is different `System.Threading.Tasks.TaskCanceledException` (forgot that websocket throws on canceled, should've checked)

up to @JustArchi whether to add catching `TaskCanceledException` exception (probably yes) and where it should be added (is there any middleware or something?)

Botan626 commented:
Now it produces errors when I just open and close ASF-ui in browser.

Updated log: https://pastebin.com/raw/v3JvHr3T

BartoszChrostowski commented:
@Botan626 You can check it now if have some time.

As I said JustArchi has to decide how to handle this exception (whether some short information should be logged), but as of now it should not log this long message.

Botan626 commented:
May I ask, did you make this long error message just disappear, or did you make kestrel to handle shutting down process without errors?

BartoszChrostowski commented:
I've added a cancellation token so that websocket is shut down gracefully.
I've also added ignoring the `TaskCanceledException`.

It is the exception that is thrown when the method finished because it was requested by cancelation so that you can differentiate between "normal" method result and the potentially malformed result (because the method didn't finish its calculations, it was requested to stop)

Ignoring this exception type is correct. *JustArchi has to decide* whether some short message that the web socket was closed due to cancellation should be displayed.

Botan626 commented:
> differentiate between "normal" method result and the potentially malformed result

In our case, when we restart ASF, the method finishes with normal or malformed result?

I tried latest build, there is no error message anymore.

BartoszChrostowski commented:
I mean... It is the WebSocket that is listening (waiting) for messages. It normally returns the message received, but when it is requested to cancel it releases resources and then returns the exception that it ended due to cancellation (before my change it encounter the server shutdown and the behavior was not so graceful because the exception was thrown during the release of the resources)

Botan626 commented:
> I mean... It is the WebSocket that is listening (waiting) for messages. It normally returns the message received, but when it is requested to cancel it releases resources and then returns the exception that it ended due to cancellation (before my change it encounter the server shutdown and the behavior was not so graceful because the exception was thrown during the release of the resources)

Thank you for explanations. So now after your changes kestrel is being told to shut down and it performs it itself instead of being shut down from outside when ASF was exiting its process?

JustArchi reviewed (approved):
I'll follow-up with some small changes but this is correct PR and seems to fix the underlying problem in the correct way, thank you a lot for your contribution! 🏆

Botan626 commented:
I gave a try to [latest ASF build](https://github.com/JustArchiNET/ArchiSteamFarm/actions/runs/425274454) and there was an error when restarting, log: https://pastebin.com/raw/ffidFkhQ

`Microsoft.AspNetCore.Connections.ConnectionAbortedException`

Botan626 commented:
Another error: https://pastebin.com/raw/M3AjS4GF

`System.OperationCanceledException`

JustArchi commented:
Post in the issue, not MR, this MR is closed.
